common/fs/src/cache/tailed_file.rs

```diff
@@ -1,109 +1,233 @@
+            if inner.offset == len {
                 return None;
             }
-        }
-
-        let mut line_groups = Vec::new();
 
-        loop {
-            let mut raw_line = Vec::new();
-            // read until a new line returning the line length
-            let line_len = match self.reader.read_until(b'\n', &mut raw_line) {
-                Ok(v) => v as u64,
-                Err(e) => {
-                    error!("error reading from file {:?}: {:?}", &paths[0], e);
-                    break;
+            // if the offset is greater than the file's len
+            // it's very likely a truncation occurred
+            if inner.offset > len {
+                info!(
+                    "{:?} was truncated from {} to {}",
+                    &paths[0], inner.offset, len
+                );
+                // Reset offset back to the start... ish?
+                // TODO: Work out the purpose of the 8192 something to do with lookback? That seems wrong.
+                inner.offset = if len < 8192 { 0 } else { len };
+                // seek to the offset, this creates the "tailing" effect
+                let offset = inner.offset;
+                if let Err(e) = inner
+                    .reader
+                    .get_mut()
+                    .get_mut()
+                    .seek(SeekFrom::Start(offset))
+                    .await
+                {
+                    error!("error seeking {:?}", e);
+                    return None;
                 }
-            };
-            // try to parse the raw data as utf8
-            // if that fails replace invalid chars with blank chars
-            // see String::from_utf8_lossy docs
-            let mut line = String::from_utf8(raw_line)
-                .unwrap_or_else(|e| String::from_utf8_lossy(e.as_bytes()).to_string());
-            // if the line doesn't end with a new line we might have read in the middle of a write
-            // so we return in this case
-            if !line.ends_with('\n') {
-                Metrics::fs().increment_partial_reads();
-                break;
             }
-            // remove the trailing new line
-            line.pop();
-            // increment the offset
-            self.offset += line_len;
-            // send the line upstream, safe to unwrap
-            debug!("tailer sendings lines for {:?}", paths);
-            line_groups.extend(paths.iter().map(|path| {
-                Metrics::fs().increment_lines();
-                Metrics::fs().add_bytes(line_len);
-                LineBuilder::new()
-                    .line(line.clone())
-                    .file(path.to_str().unwrap_or("").to_string())
-            }));
         }
 
-        if line_groups.is_empty() {
-            None
-        } else {
-            Some(futures::stream::iter(line_groups))
-        }
+        Some(
+            Lines::new(self.inner.clone())
+                .filter_map({
+                    let paths = paths.clone();
```

Comment:
The path copying is unfortunate as most lines only appear in one tracked file...